sql-plugin/src/main/scala/com/nvidia/spark/rapids/JoinGatherer.scala

```diff
@@ -670,4 +677,137 @@ case class MultiJoinGather(left: JoinGatherer, right: JoinGatherer) extends Join
   }
 
   override def toString: String = s"MULTI-GATHER $left and $right"
-}
\ No newline at end of file
+}
+
+/**
+ * JoinGatherer for existence join
+ * Existence join generates an `exists` boolean column with `true` or `false` in it,
+ * then appends it to the `output` columns. It does not shrink or expand left table.
+ * Just provides the `exists` column for the next following operator(usually filter).
+ *
+ * e.g.:
+ * <code>
+ * select * from left_table where
+ *   left_table.column_0 >= 3
+ *   or
+ *   exists (select * from right_table where left_table.column_1 < right_table.column_1)
+ *
+ * Explanation of this sql is:
+ *
+ * Filter(left_table.column_0 >= 3 or `exists`)
+ *   Existence_join // generate `exists` column, do not shrink or expand the rows of left table
+ *     left_table
+ *     right_table
+ * </code>
+ */
+class ExistenceJoinGathererImpl(
+    private val gatherMap: LazySpillableGatherMap,
+    private val data: LazySpillableColumnarBatch,
+    boundsCheckPolicy: OutOfBoundsPolicy,
+    spillCallback: Option[SpillCallback])
+    extends JoinGathererImpl(gatherMap, data, boundsCheckPolicy) {
+
+  // How much of the gather map we have output so far
+  private var gatheredUpTo: Long = 0
+
+  // Note: equals to the rows of the left table
+  private val totalRows: Long = data.numRows
+
+  override def isDone: Boolean =
+    gatheredUpTo >= totalRows
+
+  override def numRowsLeft: Long = totalRows - gatheredUpTo
+
+  override def toString: String = {
+    s"GATHERER FOR EXISTENCE JOIN $gatheredUpTo/$totalRows $gatherMap $data"
+  }
+
+  /**
+   * Gen a lazy Cb:  `left table + exists`,
+   * then close `gatherMap`` and `data`, they are useless.
+   */
+  private val lazyLeftTableWithExists: LazySpillableColumnarBatch = {
+    withResource(data) { d =>
+      val gViewTmp = withResource(gatherMap) { gMap =>
+        gMap.toColumnView(0, gMap.getRowCount.toInt)
+      }
+      val existsCvTmp = withResource(gViewTmp) { gView =>
+        genExistsColumn(gView, data.getBatch.numRows())
+      }
+      val lazyCb = withResource(existsCvTmp) { existsCv =>
+        withResource(GpuColumnVector.from(d.getBatch)) { leftTable =>
+          // add a boolean type
+          val types = GpuColumnVector.extractTypes(d.getBatch) :+ BooleanType
+          withResource(addExistsCv(leftTable, existsCv)) { tableWithExistsCv =>
+            withResource(GpuColumnVector.from(tableWithExistsCv, types)) { cb =>
+              LazySpillableColumnarBatch(cb, spillCallback.get, "existence_join_batch")
+            }
+          }
+        }
+      }
+
+      lazyCb.allowSpilling()
+      lazyCb
+    }
+  }
```

Comment:
Maybe I'm missing something, but this this looks like we're making a spillable form of the complete answer to the join batch, i.e.: the left table with the existence column.  But _we instantiated the entire result here_.  What's the point of trying to split it up afterwards?  We've already generated the entire answer here and thus already taken the memory hit, so why not emit it directly?  This is one of the reasons we did not use the `JoinGatherer` interface when implementing existence join for hash-based joins.  I'm not sure it makes a lot of sense to do so.  The gather map is not going to explode, and the column we're adding as part of the join is minimal in size (i.e.: a boolean column).

If this is going to stick with the `JoinGatherer` approach (which oddly does no such gathering in the existence join case), it seems it should not manifest the entire output up-front but only the chunks that are requested at the time they are requested.  Otherwise we can just skip the spilling overhead and just emit the answer directly (as we did in the hash-join existence case).

Reply:
Emit the batch directly.